### ISSUE TITLE
CSF story order depends on webpack and breaks with vanilla es module imports

### ISSUE THREAD

**Issue by daKmoR** [storybookjs/storybook#9136]
**Describe the bug**
The order of stories within the code of an `index.stories.js` file (in CSF format) determines the order of the storybook menu.

Consider the following example
```js
// orderedNamedExports.js
export const b = 'BB';
export const a = 'AA';
export const c = 'CC';

// start.js
import('./orderedNamedExports.js').then(module => {
  console.log(Object.keys(module));
});
```

When loading this file by going through webpack it will result in
```js
// (3) ["b", "a", "c"]
```

However, when using a browser native es module import the modules will be sorted.
```js
// (3) ["a", "b", "c"]
```

It is specified in the es module spec https://tc39.es/ecma262/#sec-modulenamespacecreate

> 7. Let sortedExports be a new List containing the same values as the list exports where the values are ordered as if an Array of the same values had been sorted using Array.prototype.sort using undefined as comparefn.

There are also part of the official platform tests https://github.com/tc39/test262/blob/master/test/language/module-code/namespace/internals/own-property-keys-sort.js

it sort of all started with this tweet so if you are interested there are more details
https://twitter.com/daKmoR/status/1204877164233015297?s=20

As this order determines the menu order it will result in "wrong menus" if you are not using webpack[1] - (and possible also when doing production builds with the upcoming es module flag in webpack v5)

**To Reproduce**
1. have a story file like
```js
import { html } from '../../index.js';

export default {
  title: 'Has order Paragraph-Heading',
};

export const paragraph = () =>
  html`
    <p>First Story</p>
  `;

export const heading = () =>
  html`
    <h1>Second Story</h1>
  `;
```
2. using webpack results in the following menu
  - Paragraph
  - Heading
3. using vanilla es module imports results in the following menu
  - Heading
  - Paragraph

**Expected behavior**
Have storybook have the same story order when using webpack and when using es modules.

**strawman proposal**
We could export a special `__exportOrder` array from within story files. This could be automatically done by a loader.

possible addition for the example above
```js
// 1.
export __exportOrder = ['paragraph', 'heading'];

// 2.
export __orderedExports = [paragraph, heading];
```

with 1. you would sort the * imports via this array and with 2. you would ignore all other imports and just use this ordered array contain all exports.

Personally I think 2. sounds better

What do you all think?

---
[1]: > A quick check shows the problem is not limited to webpack; Babel & TypeScript also make no attempt to sort exports. [tweet](https://twitter.com/robpalmer2/status/1205066610769354752?s=20)

**Comment by daKmoR** [comment]
@shilman sounds good if we prepare something like `export __orderedExports = [paragraph, heading];`?

**Comment by shilman** [comment]
How about 

```js
export const __storyOrder = ['paragraph', 'heading'];
```

**Comment by daKmoR** [comment]
Did a quick code scan and I think an ordered list of exports should be easier
probably just 1 line change? + loader of course

https://github.com/storybookjs/storybook/blob/next/lib/core/src/client/preview/start.js#L382

```diff
- const { default: meta, ...exports } = fileExports;
+ const { default: meta, __orderedExports as exports } = fileExports;
```

if we do strings then we will need to map names to exports and order them into a new object.
All possible but more extra code that can go wrong?

If you think an ordered string-list is more use useful (maybe also for other stuff) just let us know 🤗

**Comment by shilman** [comment]
@daKmoR how would one opt in to that?

**Comment by daKmoR** [comment]
I would say if a story files provides that export than it should prefer it... if not keep the same behaviour as before 

psoidocode:
```js
const { default: meta,...rawExports } = fileExports;
const exports = rawExports.__orderedExports ? rawExports.__orderedExports : rawExports;
```

a load would not even be needed for now (as it is not yet sure if webpack will change it at all and even if they will it will probably be in a breaking change)

looking at it - I think those 2 lines could be all that is needed 🤔

**Comment by shilman** [comment]
Works for me! 👍

**Comment by daKmoR** [comment]
so I tested this and it works actually really nice 👍 

one thing to mention is that the current implementation requires the key and the function... so using a simple array is not good enough => e.g. the simples would be to just use an object then 2 line changes in storybook are all that is needed

```js
// 1.
// code in story files
export const __orderedExports = { heading, card };

// storybook core
const { default: meta, __orderedExports, ...namedExports } = fileExports;
const exports = __orderedExports || namedExports;
```

if we want an array then it would need to be an array of strings which we later use to create a source ordered object out of the alphabetically ordered exports.

```js
// 2.
// code in story files
export const __storyOrder = ['heading', 'card'];

// storybook core
const { default: meta, __storyOrder, ...namedExports } = fileExports;
const exports = __storyOrder ? createExportsFromOrderAndNames(namedExports, __storyOrder) || namedExports;

function createExportsFromOrderAndNames(namedExports, order) {
  const exports = {};
  order.forEach((exportName => {
    exports[exportName] = namedExports[exportName];
  });
  return exports;
}
```

imho 1. is good enough  👍 an array seems "cleaner" but not needing to add logic to create a new object out of an array and an object seems overall preferable

sounds good? 🤗

**Comment by shilman** [comment]
@daKmoR I'm good with option 1! 💯

**Comment by Hypnosphi** [comment]
> which we later use to create a source ordered object 

You don't really need to. You can use the array for iteration, and the original exports object for values

**Comment by shilman** [comment]
Jiminy cricket!! I just released https://github.com/storybookjs/storybook/releases/tag/v5.3.0-rc.0 containing PR #9181 that references this issue. Upgrade today to try it out!

You can find this prerelease on the `@next` NPM tag.

**Comment by LarsDenBakker** [comment]
One minor thing I'm noticing while implementing this, is that perhaps we should have called it `__orderedNamedExports` since this should only contain the named exports and not the default export.

**Comment by shilman** [comment]
@LarsDenBakker This is an internal API, so happy to change it if you and @daKmoR prefer

**Comment by shilman** [comment]
Jiminy cricket!! I just released https://github.com/storybookjs/storybook/releases/tag/v5.3.0-rc.9 containing PR #9315 that references this issue. Upgrade today to try it out!

You can find this prerelease on the `@next` NPM tag.

**Comment by daKmoR** [comment]
As discussed in https://github.com/storybookjs/storybook/pull/9181 there have been 2 solutions to guarantee a story order if loaded via es modules. I went for the object. However, I have been missing the use case of rexports

```js
const myStory = () => html`Foo`;
export { myStory as notMyStory };
```

which currently produces the following `export const __orderedExports = { notMyStory };` which fails as `notMyStory` is not a local variable.

Therefore we should go with the previously discussed array of strings.

Before:
```js
export const __orderedExports = { notMyStory };
```

After:
```js
export const __namedExportsOrder = ['notMyStory'];
```

This solution has now been implemented via https://github.com/storybookjs/storybook/pull/9315
and has been released with 5.3-rc.9
